cassandra/policies.py

```diff
@@ -999,6 +999,237 @@ def shutdown(self):
+            raise ValueError("max_concurrent must be a positive integer")
+        self.backoff_policy = backoff_policy
+        self.max_concurrent = max_concurrent
+
+    def new_connection_scheduler(self, scheduler: _Scheduler) -> ShardConnectionScheduler:
+        return _LimitedConcurrencyShardConnectionScheduler(scheduler, self.backoff_policy, self.max_concurrent)
+
+
+class _ScopeBucket:
+    """
+    Holds information for a shard connection backoff policy scope, schedules and executes requests to create connection.
+    """
+    session: _Scheduler
+    backoff_policy: ShardConnectionBackoffSchedule
+    lock: Lock
+    is_shutdown: bool = False
+
+    max_concurrent: int
+    """
+    Max concurrent connection creation requests in the scope.
+    """
+
+    currently_pending: int
+    """
+    Number of currently pending connections.
+    """
+
+    items: List[Callable[[], None]]
+    """
+    List of scheduled create connections requests.
+    """
+
+    def __init__(
+            self,
+            scheduler: _Scheduler,
+            backoff_policy: ShardConnectionBackoffSchedule,
+            max_concurrent: int,
+    ):
+        self.items = []
+        self.scheduler = scheduler
+        self.backoff_policy = backoff_policy
+        self.lock = Lock()
+        self.max_concurrent = max_concurrent
+        self.currently_pending = 0
+
+    def _get_delay(self, schedule: Iterator[float]) -> Tuple[Iterator[float], float]:
+        try:
+            return schedule, next(schedule)
+        except StopIteration:
+            # A bit of trickery to avoid having lock around self.schedule
+            schedule = self.backoff_policy.new_schedule()
+            delay = next(schedule)
+            self.schedule = schedule
+            return schedule, delay
+
+    def _run(self, schedule: Iterator[float]):
+        if self.is_shutdown:
+            return
+
+        with self.lock:
+            try:
+                request = self.items.pop(0)
+            except IndexError:
+                # Just in case
+                if self.currently_pending > 0:
+                    self.currently_pending -= 1
+                # When items are exhausted reset schedule to ensure that new items going to get another schedule
+                # It is important for exponential policy
+                return
+
+        try:
+            request()
+        finally:
+            schedule, delay = self._get_delay(schedule)
+            self.scheduler.schedule(delay, self._run, schedule)
+
+    def schedule_new_connection(self, cb: Callable[[], None]):
+        with self.lock:
+            if self.is_shutdown:
+                return
+            self.items.append(cb)
+            if self.currently_pending < self.max_concurrent:
+                self.currently_pending += 1
+                schedule = self.backoff_policy.new_schedule()
+                delay = next(schedule)
+                self.scheduler.schedule(delay, self._run, schedule)
+
```

Comment:
Ok so if I understand correctly, the "concurrency" here is how many pending `scheduler.schedule` calls can there be. As far as I can tell, it doesn't do anything, since the executor is single thread.

Reply:
Ok I understand those semantics, but I don't really understand how they are useful, what is the intended use case for this? This concurrency mostly means that the sleep times will be different (because there are many "workers"), which is more difficult to reason about than different backoff_policy.

Reply:
it is useful by two reasons:
1. It gives concurrency limitation, number of worker define how many connections are handled in given time.
2. It allows to properly persist  backoff schedule, if there is no worker, we would have to either initialize it all the time, or use same backoff schedule. None of these two options is optimal: New schedule does not allow to have exponential backoff; Same schedule will produce unexpected result with exponential backoff and concurrency.